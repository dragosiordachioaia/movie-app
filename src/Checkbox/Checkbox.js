import React, { Component } from "react";

import "./Checkbox.scss";

export default class Checkbox extends Component {
  constructor() {
    super();
    this.state = {
      checked: false,
    };

    this.onClick = this.onClick.bind(this);
    this.onKeyDown = this.onKeyDown.bind(this);
    this.toggle = this.toggle.bind(this);
  }

  onClick(e) {
    if (!e.target || e.target.tagName !== "A") {
      this.toggle(e);
    }
  }

  toggle(e) {
    let newCheckedState;
    if (!this.props.hasOwnProperty("checked")) {
      newCheckedState = !this.state.checked;
      this.setState({
        checked: newCheckedState,
      });
    } else {
      newCheckedState = !this.props.checked;
    }

    if (this.props.onCheck && typeof this.props.onCheck === "function") {
      this.props.onCheck(e, newCheckedState);
    }
  }

  onKeyDown(e) {
    if (e.keyCode === 13 || e.keyCode === 32) {
      e.preventDefault();
      this.toggle(e);
    }
  }

  render() {
    let containerClassName = `checkbox ${this.props.className || ""}`;
    let checked;
    if (this.props.hasOwnProperty("checked")) {
      checked = this.props.checked;
    } else {
      checked = this.state.checked;
    }

    if (checked) {
      containerClassName += " checked";
    }

    return (
      <div onClick={this.onClick} className={containerClassName}>
        <div
          tabIndex={0}
          className="checkbox-inside"
          onKeyDown={this.onKeyDown}
        >
          <i className="fa fa-check" />
        </div>
        <label>{this.props.label}</label>
      </div>
    );
  }
}
