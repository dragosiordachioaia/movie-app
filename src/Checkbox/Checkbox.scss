@import "../variables";

.checkbox {
  display: inline-block;
  cursor: pointer;
  position: relative;

  i {
    color: $color-secondary;
    transition: all 400ms;
    transform: translate(-50%, -50%) scale(0);
    transition: all 200ms;
    transition-timing-function: ease-out;
    position: absolute;
    top: 50%;
    left: 50%;
  }

  .checkbox-inside {
    width: 1.5em;
    height: 1.5em;
    display: inline-block;
    border: 1px solid $color-secondary;
    position: relative;
    overflow: hidden;
    transition: all 500ms;
  }

  &.checked {
    i {
      transform: translate(-50%, -50%) scale(1);
    }
  }

  label {
    position: absolute;
    transform: translateY(-50%);
    left: 2.2em;
    top: 50%;
    cursor: pointer;
    user-select: none;
  }
}
