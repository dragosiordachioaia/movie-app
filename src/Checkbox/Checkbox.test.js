import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import Checkbox from "./Checkbox";

// jest.mock("Header/Header", () => "Header");

let component;
let instance;

beforeEach(() => {
  component = renderer.create(<Checkbox />);
  instance = component.getInstance();
});

it("renders without crashing", () => {
  const result = instance.render();
  expect(result).toMatchSnapshot();
});

it("renders when checked via props", () => {
  instance.props = {
    checked: true,
  };
  const result = instance.render();
  expect(result).toMatchSnapshot();
});

it("renders when unchecked via props", () => {
  instance.props = {
    checked: false,
  };
  const result = instance.render();
  expect(result).toMatchSnapshot();
});

it("does not crash if onChange callback is not a function", () => {
  instance.props = {
    checked: false,
    onChange: 3,
  };
  expect(instance.toggle).not.toThrow();
});

it("toggles on", () => {
  let mockOnCheck = jest.fn();
  instance.props = {
    checked: false,
    onCheck: mockOnCheck,
  };
  let dummyEvent = { testEvent: true };
  instance.toggle(dummyEvent);
  expect(mockOnCheck).toBeCalled();
  expect(mockOnCheck).toBeCalledWith(dummyEvent, true);
});

it("toggles off", () => {
  let mockOnCheck = jest.fn();
  instance.props = {
    checked: true,
    onCheck: mockOnCheck,
  };
  let dummyEvent = { testEvent: true };
  instance.toggle(dummyEvent);
  expect(mockOnCheck).toBeCalled();
  expect(mockOnCheck).toBeCalledWith(dummyEvent, false);
});

it("toggles on space", () => {
  let mockToggle = jest.fn();
  instance.props = {};
  instance.toggle = mockToggle;
  instance.onKeyDown({ keyCode: 13, preventDefault: jest.fn() });
  expect(mockToggle).toBeCalled();
});

it("toggles on enter", () => {
  let mockToggle = jest.fn();
  instance.props = {};
  instance.toggle = mockToggle;
  instance.onKeyDown({ keyCode: 32, preventDefault: jest.fn() });
  expect(mockToggle).toBeCalled();
});
