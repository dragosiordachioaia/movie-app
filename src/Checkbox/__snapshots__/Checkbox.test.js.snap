// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders when checked via props 1`] = `
<div
  className="checkbox  checked"
  onClick={[Function]}
>
  <div
    className="checkbox-inside"
    onKeyDown={[Function]}
    tabIndex={0}
  >
    <i
      className="fa fa-check"
    />
  </div>
  <label />
</div>
`;

exports[`renders when unchecked via props 1`] = `
<div
  className="checkbox "
  onClick={[Function]}
>
  <div
    className="checkbox-inside"
    onKeyDown={[Function]}
    tabIndex={0}
  >
    <i
      className="fa fa-check"
    />
  </div>
  <label />
</div>
`;

exports[`renders without crashing 1`] = `
<div
  className="checkbox "
  onClick={[Function]}
>
  <div
    className="checkbox-inside"
    onKeyDown={[Function]}
    tabIndex={0}
  >
    <i
      className="fa fa-check"
    />
  </div>
  <label />
</div>
`;
